package.json

```diff
@@ -1,40 +1,83 @@
 {
   "name": "foreground-child",
   "version": "2.0.0",
-  "description": "Run a child as if it's the foreground process.  Give it stdio.  Exit when it exits.",
-  "main": "index.js",
+  "description": "Run a child as if it's the foreground process. Give it stdio. Exit when it exits.",
+  "main": "./dist/cjs/index.js",
+  "module": "./dist/mjs/index.js",
+  "types": "./dist/mjs/index.d.ts",
+  "exports": {
+    ".": {
+      "import": {
+        "types": "./dist/mjs/index.d.ts",
+        "default": "./dist/mjs/index.js"
```

Comment:
Out of curiousity, how are you getting this to work without the extensions? My impression has always been that Node detects what kind of code to run based on the file type, so if `"type": "module"` is not set, then `.js` files are interpreted as CJS code, not MJS code.

Reply:
The same will apply to the `.d.ts` files too; the `.d.ts` extension always describes `.js`, which will mean they are CJS. At least, that is my understanding. (https://arethetypeswrong.github.io/ may catch this?)

Reply:
Oh, I just saw `fixup.sh`... it's putting more `package.json`s into the package, correct? Does that work? Are there published examples of this out there to test with? Genuinely curious (worried?) about what TS will do with this setup.

Reply:
Yeah, it does seem to work pretty well!  I've published a bunch of packages with this setup now.  It's a little bit of a house of cards, finding a setup that Just Works with all the different node versions, typedoc, and various webpack/rollup builders out there. I'd had a way to make default imports work smoothly across esm and cjs, but that got to be too much of a rats' nest of edge cases, so I just moved everything to exclusively using named exports.  But this exact setup does seem to work in pretty much every scenario (just means I can't use default exports).

Reply:
More examples linked on [my github.io page](https://isaacs.github.io)

Reply:
Ah, I see that rimraf's latest version does this. Seems like arethetypeswrong does figure it out, too: https://arethetypeswrong.github.io/?p=rimraf%405.0.0

Reply:
It'd be nice if you could tell TS to compile to a .mjs/.cjs without authoring it as .mts/.cts, then I could avoid having the fixup script at all, but in the meantime, just throwing a package.json file in there seems to be fine.  The other caveat besides "no default exports support" is if you want to load something like the package version from the package.json file, then ts will create `dist/cjs/package.json` and put the code in `dist/cjs/src/...`, so the fixup script has to be updated to handle that, eg: https://github.com/isaacs/rimraf/blob/main/fixup.sh

Reply:
I don't think we're ever going to do any import or extension rewriting. There are some module docs in the pipeline which I think will shed light on all of this, as it's really complicated and trying to write dual packages is fraught with gotchas and issues (e.g. how you're foregoing `default`, which is of course the main wart).

Reply:
That wart could be mostly avoided by having an option to say "no, really, set module.exports to the default export". Probably would cause an outbreak of api warts elsewhere though 😂

Reply:
Some day commonjs will die and we'll not have to worry about this. Or our grandchildren won't. If they're lucky.

Reply:
I sent this over to @andrewbranch (as this pattern isn's something I had seen before!), and he noted that you probably want to be setting `nodenext` instead of `esnext` for maximum strictness/checking.

Reply:
Also, that the top-level `"types"` should point to the CJS dir.

Reply:
> probably want to be setting nodenext instead of esnext for maximum strictness/checking.

Yes; however, the caveat is that we’re not going to notice your `{ "type": "module" }` package.json in `dist/mjs`, so `module: nodenext` will think you’re always compiling for CJS. You’d need to temporarily swap your root package.json `"type"` between the CJS and ESM compilations, which is a pain. This is something I’m hoping to address in TS soon.

Reply:
When editing, it always assumes whatever the `tsconfig.json` is (same with tests, since they get run with `ts-node`).

When I build with `nodenext` in the `tsconfig-esm.json`, it builds cjs files. If I set `"type":"module"` in the root package.json, then I also have to move `tsconfig.json` to `tsconfig-cjs.json`, `tsconfig-esm.json` to `tsconfig.json`, and write tests in mjs mode (eg, using `fileUrlToPath(import.meta.url)` instead of `__filename`).

The only difference in the built results seems to be this in all the cjs files:

```diff
@@ -1,4 +1,4 @@
-/// <reference types="node" />
+/// <reference types="node" resolution-mode="require"/>
```

Reply:
But yeah, it'd be sweet if tsc could detect the `"type"` in the nearest `package.json` file where it's putting files, or even have a way to specify that it should be built as a given type.  Then I could just place the dummy package.json files and have almost the same tsc config for both builds.